### ISSUE TITLE
Associated type in higher-rank trait bounds cause unexpected E0227 and useless help

### ISSUE THREAD

**Issue by Toru3** [rust-lang/rust#94160]
Given the following code: [code in playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=edb7c5bf275b057585f888024d52aaa5) 

```rust
use std::ops::Add;

fn add<T>(a: T, b: T) -> T
where
    T: for<'x> From<<&'x T as Add>::Output>,
    for<'x> &'x T: Add,
{
    T::from(&a + &b)
}

fn main() {
    add(1i32, 2i32); // E0227 with useless help
    //add::<i32>(1i32, 2i32); // works
}
```

The current output is:

```
error[E0277]: the trait bound `for<'x> i32: From<<&'x i32 as Add>::Output>` is not satisfied
  --> src/main.rs:12:5
   |
12 |     add(1i32, 2i32); // E0227 with useless help
   |     ^^^ the trait `for<'x> From<<&'x i32 as Add>::Output>` is not implemented for `i32`
   |
note: required by a bound in `add`
  --> src/main.rs:5:8
   |
3  | fn add<T>(a: T, b: T) -> T
   |    --- required by a bound in this
4  | where
5  |     T: for<'x> From<<&'x T as Add>::Output>,
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `add`
help: consider introducing a `where` bound, but there might be an alternative better way to express this requirement
   |
11 | fn main() where i32: for<'x> From<<&'x i32 as Add>::Output> {
   |           +++++++++++++++++++++++++++++++++++++++++++++++++

For more information about this error, try `rustc --explain E0277`.
error: could not compile `playground` due to previous error
```

Ideally the help should look like:

```
help: consider introducing the explicit type `::<i32>`
   |
12 |     add::<i32>(1i32, 2i32); // E0227 with useless help
   |        +++++++
```

Not *higher-rank* trait bounds (below code) is compiled successfully.
```rust
use std::ops::Add;

fn add<T>(a: T, b: T) -> T
where
    T: From<<T as Add>::Output> + Add,
{
    T::from(a + b)
}

fn main() {
    add(1i32, 2i32); // works
}
```

**Comment by compiler-errors** [comment]
This is fixed by #94070, and thus (likely but not checked) also fixed by #90887 (since those are basically equivalent).

> Ideally the help should look like: [...]

While this does fix the compiler error here, I am somewhat doubtful that we could make the compiler give this feedback in general. This issue is really just because the compiler has troubles with associated types that have higher-ranked lifetimes in them.

**Comment by Toru3** [comment]
If this compiles successfully, that's good enough for me. (There is no need to display help.)

**Comment by Toru3** [comment]
This issue seems same as #90729. Should I close this issue?

**Comment by compiler-errors** [comment]
I would argue that this is sufficiently different from #90729, since it doesn't require `#![feature(generic_associated_types)]`. Feel free to keep it open.

**Comment by Toru3** [comment]
I understand.

**Comment by Toru3** [comment]
FYI

```rust
use std::ops::Add;

fn ref_add<'a, T>(a: &'a T, b: &'a T) -> T
where
    T: From<<&'a T as Add>::Output>,
    &'a T: Add,
{
    T::from(a + b)
}

fn ref_add_hrtb<T>(a: &T, b: &T) -> T
where
    T: for<'x> From<<&'x T as Add>::Output>,
    for<'x> &'x T: Add,
{
    T::from(a + b)
}

fn main() {
    let a = 2i32;
    let b = 3i32;
    assert_eq!(ref_add(&a, &b), 5i32); // pass
    assert_eq!(ref_add_hrtb(&a, &b), 5i32); // E0277
}
```

[above code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7bedd965d5a59fc67bd0337f949aab6c) generate below output.

```
error[E0277]: the trait bound `for<'x> i32: From<<&'x i32 as Add>::Output>` is not satisfied
  --> src/main.rs:23:16
   |
23 |     assert_eq!(ref_add_hrtb(&a, &b), 5i32); // E0277
   |                ^^^^^^^^^^^^ the trait `for<'x> From<<&'x i32 as Add>::Output>` is not implemented for `i32`
   |
note: required by a bound in `ref_add_hrtb`
  --> src/main.rs:13:8
   |
11 | fn ref_add_hrtb<T>(a: &T, b: &T) -> T
   |    ------------ required by a bound in this
12 | where
13 |     T: for<'x> From<<&'x T as Add>::Output>,
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `ref_add_hrtb`
help: consider introducing a `where` bound, but there might be an alternative better way to express this requirement
   |
19 | fn main() where i32: for<'x> From<<&'x i32 as Add>::Output> {
   |           +++++++++++++++++++++++++++++++++++++++++++++++++

For more information about this error, try `rustc --explain E0277`.
```

**Comment by jackh726** [comment]
I think this was fixed in #90887, but needs to be confirmed.

**Comment by compiler-errors** [comment]
@jackh726, confirmed fixed on master.

**Comment by Toru3** [comment]
I confirmed fixed on nightly(1eb72580d 2022-03-08). 🎉
